File: src/main/java/top/wcpe/wcbot/WCBot.java
package top.wcpe.wcbot;

import cn.nukkit.plugin.PluginBase;
import cn.nukkit.utils.ConfigSection;
import lombok.Getter;
import net.mamoe.mirai.contact.Friend;
import net.mamoe.mirai.contact.Group;
import top.wcpe.wcbot.botcommand.command.*;
import top.wcpe.wcbot.inter.DataManager;
import top.wcpe.wcbot.inter.impl.DataManagerFileDataImpl;
import top.wcpe.wcbot.window.EventWindow;
import top.wcpe.wcpelib.nukkit.command.CommandPlus;
import top.wcpe.wcpelib.nukkit.command.entity.Command;
import top.wcpe.wcpelib.nukkit.command.entity.CommandArgument;
import top.wcpe.wcpelib.nukkit.manager.MessageManager;

/**
 * 插件主类
 *
 * @author: WCPE 1837019522@qq.com
 * @create: 2021-06-22 14:09
 */
public class WCBot extends PluginBase {


    @Getter
    private static WCBot instance;

    @Getter
    private static MessageManager messageManager;
    @Getter
    private static DataManager dataManager;
    @Getter
    private static BotManager botManager;

    @Getter
    private static WcBotBasicFunctions wcBotBasicFunctions;


    public ConfigSection getSetting() {
        return getConfig().getSection("Setting");
    }

    @Override
    public void onEnable() {
        long start = System.currentTimeMillis();
        instance = this;
        saveDefaultConfig();
        messageManager = new MessageManager(this, "CN");

        dataManager = new DataManagerFileDataImpl();
        botManager = new BotManager();

        wcBotBasicFunctions = new WcBotBasicFunctions();

        getServer().getPluginManager().registerEvents(wcBotBasicFunctions, this);

        CommandPlus cp = new CommandPlus.Builder("WC-Bot", this).aliases("wb").build();
        cp.registerSubCommand(new Command.Builder("sendGroup", "§a发送消息至QQ群").executeComponent((sender, args) -> {
            Long qqGroupId = null;
            try {
                qqGroupId = Long.parseLong(args[0]);
            } catch (Exception e) {
                sender.sendMessage("请输入正确的QQ群号码!");
                return;
            }
            Group group = getBotManager().getDefaultBot().getGroup(qqGroupId);
            if (group == null) {
                sender.sendMessage("请输入正确的QQ群号码!");
                return;
            }
            group.sendMessage(args[1]);
            sender.sendMessage("发送成功!");
        }).args(new CommandArgument.Builder("QQ群号").build(), new CommandArgument.Builder("消息").build()).build());
        cp.registerSubCommand(new Command.Builder("sendQQ", "§a发送消息至QQ").executeComponent((sender, args) -> {
            Long qqFirendId = null;
            try {
                qqFirendId = Long.parseLong(args[0]);
            } catch (Exception e) {
                sender.sendMessage("请输入正确的QQ号码!");
                return;
            }
            Friend friend = getBotManager().getDefaultBot().getFriend(qqFirendId);
            if (friend == null) {
                sender.sendMessage("请输入正确的QQ号码!");
                return;
            }
            friend.sendMessage(args[1]);
            sender.sendMessage("发送成功!");
        }).args(new CommandArgument.Builder("QQ号").build(), new CommandArgument.Builder("消息").build()).build());


        cp.registerThis();
        new AdminManagerCommand().registerCommand();
        new BindManagerCommand().registerCommand();
        new BotPointManagerCommand().registerCommand();
        new EnableGroupManagerCommand().registerCommand();
        new ListCommand().registerCommand();
        new PlayerBindCommand().registerCommand();
        new PlayerGameTimeCommand().registerCommand();
        new PlayerInfoCommand().registerCommand();
        new PlayerSignCommand().registerCommand();
        new SudoCommand().registerCommand();
        new WhiteListCommand().registerCommand();
        new WhiteListManagerCommand().registerCommand();

        getLogger().info("load time: " + (System.currentTimeMillis() - start) + " ms");
        getServer().getConsoleSender().sendMessage("§1  _       __   ______           ____           __ ");
        getServer().getConsoleSender().sendMessage("§2 | |     / /  / ____/          / __ )  ____   / /_");
        getServer().getConsoleSender().sendMessage("§3 | | /| / /  / /      ______  / __  | / __ \\ / __/");
        getServer().getConsoleSender().sendMessage("§4 | |/ |/ /  / /___   /_____/ / /_/ / / /_/ // /_  ");
        getServer().getConsoleSender().sendMessage("§5 |__/|__/   \\____/          /_____/  \\____/ \\__/  ");
        getServer().getConsoleSender().sendMessage("§6                                                  ");
        getServer().getConsoleSender().sendMessage("§6              §aServerVersion: " + getServer().getVersion() + "      ");
        getServer().getConsoleSender().sendMessage("§6              §aPluginVersion: 3.0.0      ");
        getServer().getConsoleSender().sendMessage("§a如果您喜欢这个插件 请在 Mcbbs 帖子: https://www.mcbbs.net/thread-971000-1-1.html 下方给我评个分~");
        new bStats(this, 6812);
        getLogger().info("onEnable...");

        if (getConfig().getBoolean("EventWindow.enable"))
            new EventWindow();
    }


    @Override
    public void onDisable() {
        getLogger().info("onDisable...");
    }
}

File: src/main/kotlin/top/wcpe/wcbot/WCBot.kt
package top.wcpe.wcbot

import cn.nukkit.plugin.PluginBase
import cn.nukkit.utils.Config
import top.wcpe.wcbot.bot.BotCommands
import top.wcpe.wcbot.bot.BotListener
import top.wcpe.wcbot.bot.BotManager
import top.wcpe.wcbot.data.FileDataManager
import top.wcpe.wcbot.data.IDataManager
import top.wcpe.wcbot.entity.ServerData
import top.wcpe.wcbot.nukkit.NukkitCommands
import top.wcpe.wcbot.nukkit.NukkitListener
import top.wcpe.wcbot.nukkit.NukkitTask
import top.wcpe.wcpelib.nukkit.manager.MessageManager
import java.nio.file.Files
import java.nio.file.Path

/**
 * 由 WCPE 在 2021/11/24 20:14 创建
 * Created by WCPE on 2021/11/24 20:14
 * Github: https://github.com/wcpe
 * QQ: 1837019522
 *
 * @author WCPE
 */
class WCBot : PluginBase() {

    companion object {

        @JvmStatic
        lateinit var instance: WCBot

        @JvmStatic
        lateinit var messageManager: MessageManager

        @JvmStatic
        lateinit var dataManager: IDataManager

        @JvmStatic
        lateinit var botManager: BotManager

        @JvmStatic
        lateinit var serverData: ServerData

    }


    private lateinit var configPath: Path
    private lateinit var config: Config
    private lateinit var customReplyPath: Path
    lateinit var customReplyConfig: Config
    private lateinit var groupsPath: Path
    lateinit var groupsConfig: Config
    private lateinit var commandsPath: Path
    lateinit var commandsConfig: Config

    override fun saveDefaultConfig() {
        if (Files.notExists(configPath)) {
            this.saveResource("config.yml", false)
        }
        if (Files.notExists(customReplyPath)) {
            this.saveResource("customReply.yml", false)
        }
        if (Files.notExists(groupsPath)) {
            this.saveResource("groups.yml", false)
        }
        if (Files.notExists(commandsPath)) {
            this.saveResource("commands.yml", false)
        }
    }

    override fun getConfig(): Config {
        return config
    }

    override fun reloadConfig() {
        config.reload()
        groupsConfig.reload()
        customReplyConfig.reload()
        commandsConfig.reload()
    }

    private fun initConfig() {
        configPath = dataFolder.toPath().resolve("config.yml")
        customReplyPath = dataFolder.toPath().resolve("customReply.yml")
        groupsPath = dataFolder.toPath().resolve("groups.yml")
        commandsPath = dataFolder.toPath().resolve("commands.yml")
        saveDefaultConfig()
        config = Config(configPath.toFile())
        customReplyConfig = Config(customReplyPath.toFile())
        groupsConfig = Config(groupsPath.toFile())
        commandsConfig = Config(commandsPath.toFile())
    }

    lateinit var botCommands: BotCommands
    lateinit var botListener: BotListener
    lateinit var nukkitCommands: NukkitCommands
    lateinit var hookOtherPlugin: HookOtherPlugin

    override fun onEnable() {
        val start = System.currentTimeMillis()
        logger.info("onEnable...")
        instance = this
        logger.info("初始化各项配置!")
        initConfig()
        logger.info("读取服务器数据!")
        serverData = ServerData()
        logger.info("创建消息管理器!")
        messageManager = MessageManager(this, "CN")
        logger.info("创建数据管理器!")
        dataManager = FileDataManager()
        logger.info("注册机器人管理器!")
        botManager = BotManager()
        logger.info("注册机器人指令!")
        botCommands = BotCommands()
        logger.info("注册机器人监听器!")
        botListener = BotListener()
        logger.info("注册 Nukkit 指令!")
        nukkitCommands = NukkitCommands()
        logger.info("注册 Nukkit 任务!")
        NukkitTask()
        logger.info("注册 Nukkit 监听器!")
        server.pluginManager.registerEvents(NukkitListener(), this)
        logger.info("启动 BStats!")
        bStats(this, 6812)

        logger.info("开始 Hook 其他插件!")
        hookOtherPlugin = HookOtherPlugin()

        server.consoleSender.sendMessage("§1  _       __   ______           ____           __ ")
        server.consoleSender.sendMessage("§2 | |     / /  / ____/          / __ )  ____   / /_")
        server.consoleSender.sendMessage("§3 | | /| / /  / /      ______  / __  | / __ \\ / __/")
        server.consoleSender.sendMessage("§4 | |/ |/ /  / /___   /_____/ / /_/ / / /_/ // /_  ")
        server.consoleSender.sendMessage("§5 |__/|__/   \\____/          /_____/  \\____/ \\__/  ")
        server.consoleSender.sendMessage("§6                                                  ")
        server.consoleSender.sendMessage("§6              §aServerVersion: " + server.version + "      ")
        server.consoleSender.sendMessage("§6              §aPluginVersion: ${description.version}      ")
        server.consoleSender.sendMessage("§a如果您喜欢这个插件 请在 Mcbbs 帖子: https://www.mcbbs.net/thread-971000-1-1.html 下方给我评个分~")
        logger.info("load time: " + (System.currentTimeMillis() - start) + " ms")
    }


    override fun onDisable() {
        logger.info("onDisable...")
    }
}